xcm/src/v0/multi_location.rs

```diff
@@ -44,77 +44,423 @@ use parity_scale_codec::{self, Decode, Encode};
+					} else {
+						// a single byte may not be enough to decode a `Junction`, so keep reading
+						// the next byte until a `Junction` can successfully be decoded
+						bytes.push(input.read_byte()?);
+					}
+				}
+			}
+		}
+		Ok(MultiLocation { parents, interior })
+	}
 }
 
 /// Maximum number of junctions a `MultiLocation` can contain.
-pub const MAX_MULTILOCATION_LENGTH: usize = 8;
+pub const MAX_MULTILOCATION_LENGTH: usize = 255;
+
+impl MultiLocation {
+	/// Creates a new `MultiLocation`, ensuring that the length of it does not exceed the maximum,
+	/// otherwise returns `Err`.
+	pub fn new(parents: u8, junctions: Junctions) -> result::Result<MultiLocation, ()> {
+		if parents as usize + junctions.len() > MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		Ok(MultiLocation { parents, interior: junctions })
+	}
+
+	/// Creates a new `MultiLocation` with 0 parents and a `Null` interior.
+	pub const fn empty() -> MultiLocation {
+		MultiLocation { parents: 0, interior: Junctions::Null }
+	}
+
+	/// Creates a new `MultiLocation` with the specified number of parents in the `P` const generic
+	/// parameter and a `Null` interior.
+	// #FIXME: Use a where clause to evaluate P for well-formedness once const eval of const
+	//         generics is possible.
+	pub const fn with_parents_const<const P: u8>() -> MultiLocation {
+		MultiLocation { parents: P, interior: Junctions::Null }
+	}
+
+	/// Creates a new `MultiLocation` with the specified number of parents and a `Null` interior.
+	/// Returns an error if `parents` is greater than `MAX_MULTILOCATION_LENGTH`.
+	pub const fn with_parents(parents: u8) -> result::Result<MultiLocation, ()> {
+		if parents as usize > MAX_MULTILOCATION_LENGTH {
+			return Err(())
+		}
+		Ok(MultiLocation { parents, interior: Junctions::Null })
+	}
+
+	/// Creates a new `MultiLocation` with no parents and a single `Parachain` interior junction
+	/// specified by `para_id`.
+	pub const fn with_parachain_interior(para_id: u32) -> MultiLocation {
+		MultiLocation { parents: 0, interior: Junctions::X1(Junction::Parachain(para_id)) }
+	}
+
+	/// Whether or not the `MultiLocation` has no parents and has a `Null` interior.
+	pub const fn is_empty(&self) -> bool {
+		self.parents == 0 && self.interior.len() == 0
+	}
+
+	/// Return a reference to the interior field.
+	pub fn interior(&self) -> &Junctions {
+		&self.interior
+	}
+
+	/// Return a mutable reference to the interior field.
+	pub fn interior_mut(&mut self) -> &mut Junctions {
+		&mut self.interior
+	}
+
+	/// Returns the number of `Parent` junctions at the beginning of `self`.
+	pub const fn parent_count(&self) -> u8 {
+		self.parents
+	}
+
+	/// Returns the number of parents and junctions in `self`.
+	pub const fn len(&self) -> usize {
+		self.parent_count() as usize + self.interior.len()
+	}
+
+	/// Returns the first interior junction, or `None` if the location is empty or contains only
+	/// parents.
+	pub fn first_interior(&self) -> Option<&Junction> {
```

Comment:
can't you just do `foo.iterior().first()`?

Reply:
Creating a new iterator just to grab the first element seems a bit inefficient to me; `self.interior.first()` is more direct and creates less stuff on the stack.

Reply:
sorry what I mean was `foo.interior().first()`